### PR TITLE
New handler signature

### PR DESCRIPTION
**Pull request by JockLawrie**
Tests for existing signature pass.
Tests for new signature pass.
However, I'm having trouble using the do syntax.

### CODE REVIEW

**Comment by JockLawrie** [comment]
Build error has something to do with Homebrew. Any ideas? I'm new to this.

**Comment by malmaud** [comment]
That one isn't your fault, home brew is broke at the moment.

**Comment by malmaud** [comment]
i reran the tests now that hb is fixed

**Comment by JockLawrie** [comment]
So we're good to go?

File: src/HttpServer.jl [hunk 1]
```diff
@@ -89,14 +89,24 @@ handler.events["foo"] = (bar) "Hello \$bar"
...
+
 immutable HttpHandler
     handle::Function
     sock::Base.TCPServer
     events::Dict
 
-    HttpHandler(handle::Function, sock::Base.TCPServer) = new(handle, sock, defaultevents)
-    HttpHandler(handle::Function) = new(handle, Base.TCPServer(), defaultevents)
+    function HttpHandler(handle::Function, sock::Base.TCPServer)
+        try
```
Comment (malmaud) [inline src/HttpServer.jl]: It's possible that the user passes in a generic function to do the HttpHandler constructor, which at the time doesn't have a two-argument method defined, but the user might define a two-argument method on that function  before the server is actually run. So this code's logic isn't perfectly backwards-compatible.

**Comment by malmaud** [comment]
I think a more robust approach that doesn't involve try/catch blocks (which is a bit hacky) is to use the `method_exists` function to check if a one-argument version of the handler exists *right before it's called* in the course of serving a request, rather than in the HttpHandler constructor. ie, call `method_exists(handle, (Request,))` to determine if the one-argument or two-argument version should be used, raising a deprecation warning in the second case. 

Does that make sense?

**Comment by JockLawrie** [comment]
Yes that makes sense. I'll update the PR on Tuesday (I can't get to it until then).

**Comment by malmaud** [comment]
No rush. Would also be nice if you could do a quick benchmark with say https://httpd.apache.org/docs/2.2/programs/ab.html to see if there's noticable regression in the two-argument case introduced by this PR - it wouldn't be great if the existing users who are using the old two-argument form suddenly saw a performance drop.

**Comment by JockLawrie** [comment]
Indeed. Will do. Thanks for the suggestions.

While we're at it, I can benchmark locally for my own inner peace but others won't be able to reproduce the results. Any suggestions for having benchmarking as part of this package?

**Comment by JockLawrie** [comment]
I suppose at a minimum, include an example URLs and a benchmarking script. Then users can benchmark the url locally using different commits. Fiddly, but a start.

**Comment by JockLawrie** [comment]
``method_exists`` doesn't work for anonymous functions, i.e., do syntax. So for now I've stayed with try/catch which is admittedly hacky.

**Comment by JockLawrie** [comment]
@malmaud Would you prefer to warn when processing each request or only when processing the first request?

**Comment by JockLawrie** [comment]
Added a simple benchmark to the examples folder.

On my machine the benchmark for the 2 argument handler takes about 2.9 seconds, whereas the benchmark for the 1 argument handler takes about 2.7 seconds. Not sure if this is a genuine difference or simply related to printing the warning.

File: src/HttpServer.jl [hunk 2]
```diff
@@ -89,15 +89,29 @@ handler.events["foo"] = (bar) "Hello \$bar"
...
-handle(handler::HttpHandler, req::Request, res::Response) = handler.handle(req, res)
+HttpHandler(handle::Function) = HttpHandler(handle, Base.TCPServer())
+
+function handle(handler::HttpHandler, req::Request, res::Response)
+    # Uses try/catch because method_exists doesn't work for anonymous functions...i.e., do syntax.
+    try
+        @which handler.handle(Request())    # Check if signature of handle is (:Request) 
+	handler.handle(req)
+    catch err
+        Base.warn_once("The use of handler(Request, Response) is deprecated. Please use handler(Request) instead.")
```
Comment (malmaud) [inline src/HttpServer.jl]: This will also catch errors thrown by the handler, even if the one argument handler does exist

**Comment by JockLawrie** [comment]
Can this go ahead?

**Pull request closed without merging**